### PR TITLE
Linear Algebra factory mods

### PR DESCRIPTION
**Pull request by cnpetra**
Linear algebra factory can allocate linear algebra objects in all supported memory space dynamically (at runtime). 

The PR addresses https://github.com/LLNL/hiop/issues/131.

A couple of linear algebra classes were instrumented with `mem_space_` members to keep track of the memory space. Other than that, the changes are just due to change in the signature of the `create_xxx` methods of the linear algebra factory.  

It looks like the changes for supporting multiple execution policies at runtime can be localized to the linear algebra factory.

### CODE REVIEW

**Comment by ashermancinelli** [comment]
I think this is great!

File: src/Optimization/hiopAlgPrimalDecomp.cpp [hunk 1]
```diff
@@ -127,8 +132,7 @@ hiopAlgPrimalDecomposition::HessianApprox::
 }
 
 /* n_ is the dimension of x, hence the dimension of g_k, skm1, etc */
-void hiopAlgPrimalDecomposition::HessianApprox::
-set_n(const int n)
+void hiopAlgPrimalDecomposition::HessianApprox::set_n(const int n)
 {
   n_=n;
```
Comment (nychiang) [inline src/Optimization/hiopAlgPrimalDecomp.cpp:137]:
```suggestion
  n_ = n;
```

File: src/Optimization/hiopAlgPrimalDecomp.hpp [hunk 2]
```diff
@@ -242,7 +243,7 @@ class hiopAlgPrimalDecomposition
   //indices of the coupled x in the full x
   int* xc_idx_;
   //tolerance of the convergence stopping criteria
```
Comment (nychiang) [inline src/Optimization/hiopAlgPrimalDecomp.hpp:245]:
```suggestion
  //tolerance of the convergence stopping criteria. TODO: use user option
```

File: src/Optimization/hiopNlpFormulation.cpp [hunk 3]
```diff
@@ -914,7 +927,7 @@ bool hiopNlpFormulation::user_callback_iterate(int iter,
   return interface_base.iterate_callback(iter, obj_value, logbar_obj_value,
 					 (int)n_vars, x.local_data_const(),
 					 z_L.local_data_const(), z_U.local_data_const(),
-           (int)n_cons_ineq, s.local_data_const(),
+                                         (int)n_cons_ineq, s.local_data_const(),
 					 (int)n_cons, cons_body_->local_data_const(),
 					 cons_lambdas_->local_data_const(),
 					 inf_pr, inf_du, onenorm_pr, mu, alpha_du, alpha_pr,  ls_trials);
```
Comment (nychiang) [inline src/Optimization/hiopNlpFormulation.cpp:933]: indention

**Review by nychiang — APPROVED** [review]
LGTM.
Before closing this PR, it's better for us to file a new issue to mention the temporary fix of `mem_space_` in `hiopInterfacePrimalDecomp`

**Pull request merged**